marklogic-client-api/src/main/java/com/marklogic/client/datamovement/ZipSplitter.java

```diff
@@ -0,0 +1,363 @@
+        private ZipEntry zipEntry;
+        private Format   format;
+
+        private ZipEntry getZipEntry() {
+            return this.zipEntry;
+        }
+
+        private void setZipEntry(ZipEntry zipEntry) {
+            if (zipEntry == null) {
+                throw new IllegalArgumentException("ZipEntry cannot be null");
+            }
+
+            this.zipEntry = zipEntry;
+        }
+
+        private Format getFormat() {
+            return this.format;
+        }
+
+        private void setFormat(Format format) {
+            if (format == null) {
+                throw new IllegalArgumentException("Format cannot be null");
+            }
+
+            this.format = format;
+        }
+    }
+
+    private static abstract class ZipEntrySpliterator<T> extends Spliterators.AbstractSpliterator<T> {
+        private static Pattern extensionRegex = Pattern.compile("\\.([^.]+)$");
+        private ZipInputStream zipStream;
+        private Map<String,Format> extensionFormats;
+        private Predicate<ZipEntry> entryFilter;
+        private Function<String, String> uriTransformer;
+
+        ZipEntrySpliterator(long est, int additionalCharacteristics) {
+            super(est, additionalCharacteristics);
+        }
+
+        ZipInputStream getZipStream() {
+            return this.zipStream;
+        }
+
+        void setZipStream(ZipInputStream zipStream) {
+            if (zipStream == null) {
+                throw new IllegalArgumentException("ZipStream cannot be null");
+            }
+
+            this.zipStream = zipStream;
+        }
+
+        Map<String, Format> getExtensionFormats() {
+            return this.extensionFormats;
+        }
+
+        void setExtensionFormats(Map<String, Format> extensionFormats) {
+            if (extensionFormats == null) {
+                throw new IllegalArgumentException("ExtensionFormats cannot be null");
+            }
+            this.extensionFormats = extensionFormats;
+        }
+
+        Predicate<ZipEntry> getEntryFilter() {
+            return this.entryFilter;
+        }
+
+        void setEntryFilter(Predicate<ZipEntry> entryFilter) {
+            this.entryFilter = entryFilter;
+        }
+
+        Function<String, String> getUriTransformer() {
+            return this.uriTransformer;
+        }
+
+        void setUriTransformer(Function<String, String> uriTransformer) {
+            this.uriTransformer = uriTransformer;
+        }
+
+        protected FormatEntry getNextEntry() throws IOException {
+            ZipEntry candidateEntry;
+
+            while ((candidateEntry = getZipStream().getNextEntry()) != null) {
+                if (getEntryFilter() != null && getEntryFilter().test(candidateEntry) == false) {
+                    continue;
+                }
+
+                String name = candidateEntry.getName();
+                if (getUriTransformer() != null) {
+                    name = getUriTransformer().apply(name);
+                }
```

Comment:
My mistake.  DocumentWriteOperationSpliterator.tryAdvance() calls the URI transformer in the correct place.

I don't think we should call URI transformer here -- that would be calling it twice on the same name.

Maybe the uriTransformer field and its getter and setter methods should move from the base ZipEntrySpliterator class to the derived DocumentWriteOperationSpliterator class.  If so, we should only initialize the DocumentWriteOperationSpliterator instance with the uriTransformer and not the BytesHandleSpliterator instance.

Apologies for the mistaken suggestion during the working session.

Reply:
Removed getUriTransformer call in getNextEntry method.
Removed setUriTransfomer in split method as bytesHandleSpliterator doesn't need it.
Moved getter and setter f uriTransformer field from base ZipEntrySpliterator to DcoumentWriteOperationSplitorator.

Reply:
Actually we don't need the getter and setter in DcoumentWriteOperationSplitorator. Because we pass in ZipSplitter splitter when constructing DocumentWriteOperationSpliterator, we could call splitter.uriTransformer directly. What do you think?

Reply:
Sounds good to me.